### PR TITLE
Remove deps

### PR DESCRIPTION
**Pull request by Jarcho**
This remove both `regex` and `cargo_metadata` as dependencies making `clippy_dev` compile ~3x faster (~46s -> ~16s locally). `cargo_metadata` was used to extract the `version` field from `Cargo.toml`, which is done trivially without that. `regex` was used to parse `define_clippy_lint` in `update_lints` which is now done using `rustc_lexer`. This isn't any simpler, but it compiles ~15s faster and runs ~3x faster (~2.1s -> ~0.7s locally).

The next biggest offenders to compile times are `clap` and `winapi` on windows. `clap` could be removed, but re-implementing enough is probably more work than it's worth. `winapi` is used by `opener` and `walkdir` so it's stuck there.

changelog: none

### CODE REVIEW

**Review by xFrednet — COMMENTED** [review]
This is a solid improvement and excellent work. I have one comment and some nits, but just little things.

The changes also had a major improvement for my compilation performance, it cut 30 seconds down to ~9 seconds. And the code is also more readable, IMO. Okay, enough praise :upside_down_face:

File: clippy_dev/src/update_lints.rs [hunk 1]
```diff
@@ -1,46 +1,18 @@
...
 use std::ffi::OsStr;
 use std::fs;
-use std::lazy::SyncLazy;
 use std::path::Path;
 use walkdir::WalkDir;
 
 use crate::clippy_project_root;
 
-const GENERATED_FILE_COMMENT: &str = "// This file was generated by `cargo dev update_lints`.\n\
+static GENERATED_FILE_COMMENT: &str = "// This file was generated by `cargo dev update_lints`.\n\
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_dev/src/update_lints.rs:12]: I'm not sure why that changed. Probably happened while editing things by accident.

File: clippy_dev/src/update_lints.rs [hunk 2]
```diff
@@ -330,61 +276,123 @@ fn gen_register_lint_list<'a>(
...
+            .by_ref()
+            .filter(|&(kind, _)| !matches!(kind, TokenKind::Whitespace | TokenKind::LineComment { .. }));
+        let (name, reason) = match_tokens!(
+            iter,
+            Bang OpenBrace
+            Pound OpenBracket Ident Colon Colon Ident Eq Literal{..} CloseBracket
+            Ident Ident(name) Comma
+            Literal{kind: LiteralKind::Str{..},..}(reason)
+            CloseBrace
+        );
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_dev/src/update_lints.rs:382]: I had considered making the macro use actual tokens, but it wasn't worth it just to be used twice. I'll add the comments in.

File: clippy_dev/src/update_lints.rs [hunk 3]
```diff
@@ -60,60 +32,52 @@ pub enum UpdateMode {
...
-    )
-    .changed;
+        |res| {
+            for lint in usable_lints
+                .iter()
+                .map(|l| &l.name)
+                .chain(deprecated_lints.iter().map(|l| &l.name))
+                .sorted()
+            {
+                let _ = writeln!(res, "[`{}`]: {}#{}", lint, DOCS_LINK, lint);
```
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_dev/src/update_lints.rs:64]: Both are writing to a `String` so they can't fail. There wouldn't be anything wrong with `unwrap` here, but it doesn't add much.

File: clippy_dev/src/update_lints.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_dev/src/update_lints.rs]: Both are now const.

File: clippy_dev/src/update_lints.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jarcho) [reply clippy_dev/src/update_lints.rs]: `unwrap` added just in case someone comes and changes the type to something else.

**Comment by Jarcho** [comment]
The win isn't as big in practice as `cargo_metadata` is still used in `clippy_lints`, but `regex` is completely removed.

**Review by xFrednet — APPROVED** [review]

**Comment by xFrednet** [comment]
Everything looks good to me, thank you for this change. I think everyone is going to enjoy the faster compile time for `clippy_dev` :tada: 

@bors r+

**Pull request merged**